src/MatrixFields/field_matrix_iterative_solver.jl

```diff
@@ -0,0 +1,276 @@
+has_lazy_inverse(P_alg) = isnothing(P_alg) || isnothing(solver_algorithm(P_alg))
+
+lazy_preconditioner(::Nothing, A) = one(A)
+
+function preconditioner_cache(P_alg, A, b)
+    has_lazy_inverse(P_alg) && return (;)
+    lazy_P = lazy_preconditioner(P_alg, A)
+    is_lazy_P_concrete = !(lazy_P isa FieldMatrixBroadcasted)
+    P = is_lazy_P_concrete ? lazy_P : Base.Broadcast.materialize(lazy_P)
+    P_if_needed = is_lazy_P_concrete ? (;) : (; P)
+    x = similar_to_x(P, b)
+    alg = solver_algorithm(P_alg)
+    cache = field_matrix_solver_cache(alg, P, b)
+    return (; P_if_needed..., b = similar(b), x, cache)
+end
+
+function check_preconditioner(P_alg, P_cache, A, b)
+    isnothing(P_alg) && return
+    lazy_P = lazy_preconditioner(P_alg, A)
+    if has_lazy_inverse(P_alg)
+        check_block_diagonal_matrix_has_no_missing_blocks(lazy_P, b)
+    else
+        alg = solver_algorithm(P_alg)
+        check_field_matrix_solver(alg, P_cache.cache, lazy_P, b)
+    end
+end
+
+function preconditioner(P_alg, P_cache, A)
+    isnothing(P_alg) && return nothing
+    lazy_P = lazy_preconditioner(P_alg, A)
+    is_lazy_P_concrete = !(lazy_P isa FieldMatrixBroadcasted)
+    (has_lazy_inverse(P_alg) || is_lazy_P_concrete) && return lazy_P
+    @. P_cache.P = lazy_P
+    return P_cache.P
+end
+
+function apply_preconditioner(P_alg, P_cache, P, lazy_b)
+    isnothing(P_alg) && return lazy_b
+    has_lazy_inverse(P_alg) && return lazy_mul(lazy_inv(P), lazy_b)
+    @. P_cache.b = lazy_b
+    alg = solver_algorithm(P_alg)
+    field_matrix_solve!(alg, P_cache.cache, P_cache.x, P, P_cache.b)
+    return P_cache.x
+end
+
+################################################################################
+
+abstract type PreconditionerAlgorithm end
```

Comment:
what is the interface a `PreconditionerAlgorithm` needs to define?

Reply:
Each preconditioner algorithm needs to define `lazy_preconditioner`, which generates an un-materialized form of the preconditioner matrix, and `solver_algorithm`, which specifies how to invert the preconditioner matrix.